lib/klaviyo/client.rb

```diff
@@ -1,74 +1,97 @@
-require 'open-uri'
-require 'base64'
-require 'json'
-
 module Klaviyo
-  class KlaviyoError < StandardError; end
-  
   class Client
-    def initialize(api_key, url = 'http://a.klaviyo.com/')
-      @api_key = api_key
-      @url = url
-    end
-    
-    def track(event, kwargs = {})
-      defaults = {:id => nil, :email => nil, :properties => {}, :customer_properties => {}, :time => nil}
-      kwargs = defaults.merge(kwargs)
-      
-      if kwargs[:email].to_s.empty? and kwargs[:id].to_s.empty?
-        raise KlaviyoError.new('You must identify a user by email or ID')
+    BASE_API_URL = 'https://a.klaviyo.com/api'
+    V1_API = 'v1'
+    V2_API = 'v2'
+
+    HTTP_DELETE = 'delete'
+    HTTP_GET = 'get'
+    HTTP_POST = 'post'
+    HTTP_PUT = 'put'
+
+    ALL = 'all'
+    TIMELINE = 'timeline'
+
+    DEFAULT_COUNT = 100
+    DEFAULT_PAGE = 0
+    DEFAULT_SORT_DESC = 'desc'
+
+    private
+
+    def self.request(method, path, kwargs = {})
+      check_private_api_key_exists()
+      url = "#{BASE_API_URL}/#{path}"
+      connection = Faraday.new(
+        url: url,
+        headers: {
+          'Content-Type' => 'application/json'
+      })
+      response = connection.send(method) do |req|
+        req.body = kwargs[:body].to_json || nil
       end
-      
-      customer_properties = kwargs[:customer_properties]
-      customer_properties[:email] = kwargs[:email] unless kwargs[:email].to_s.empty?
-      customer_properties[:id] = kwargs[:id] unless kwargs[:id].to_s.empty?
+      puts response.body
+    end
+
+    def self.public_request(method, path, kwargs = {})
+      check_public_api_key_exists()
+      params = build_params(kwargs)
+      url = "#{BASE_API_URL}/#{path}?#{params}"
+      res = Faraday.get(url).body
+    end
 
-      params = {
-        :token => @api_key,
-        :event => event,
-        :properties => kwargs[:properties],
-        :customer_properties => customer_properties,
-        :ip => ''
+    def self.v1_request(method, path, kwargs = {})
+      defaults = {:page => nil,
+                  :count => nil,
+                  :since => nil,
+                  :sort => nil}
+      params = defaults.merge(kwargs)
+      query_params = encode_params(params)
+      url_params = "api_key=#{Klaviyo.private_api_key}#{query_params}"
```

Comment:
i think you could just do `url = "#{V1_API}/#{path}?api_key=#{Klaviyo.private_api_key}#{query_params}"`